### PR TITLE
chore(protocol): do client hello instead of server hello

### PR DESCRIPTION
**Pull request by mxschmitt**
Relates #7741
Python implementation https://github.com/microsoft/playwright-python/pull/836

### CODE REVIEW

File: src/dispatchers/dispatcher.ts [hunk 1]
```diff
@@ -125,6 +126,12 @@ class Root extends Dispatcher<{ guid: '' }, {}> {
   constructor(connection: DispatcherConnection) {
     super(connection, { guid: '' }, '', {}, true);
   }
+
+  initialize(params: { language?: string }): import('./playwrightDispatcher').PlaywrightDispatcher {
```
Comment (kumaraditya303) [inline src/dispatchers/dispatcher.ts:130]: You can use `import type` here

File: src/dispatchers/dispatcher.ts [hunk 2]
```diff
@@ -121,15 +122,21 @@ export class Dispatcher<Type extends { guid: string }, Initializer> extends Even
...
 export type DispatcherScope = Dispatcher<any, any>;
-class Root extends Dispatcher<{ guid: '' }, {}> {
-  constructor(connection: DispatcherConnection) {
+export class Root extends Dispatcher<{ guid: '' }, {}> {
+  constructor(connection: DispatcherConnection, private readonly createPlaywright?: (scope: DispatcherScope) => Promise<PlaywrightDispatcher>) {
     super(connection, { guid: '' }, '', {}, true);
   }
+
+  async initialize(params: { language?: string }): Promise<channels.RootInitializeResult> {
+    assert(this.createPlaywright);
```
Comment (mxschmitt) [inline src/dispatchers/dispatcher.ts:131]: note: for inprocess we don't need it, for everything else we use it.

File: src/dispatchers/dispatcher.ts [hunk 3]
```diff
@@ -121,15 +122,21 @@ export class Dispatcher<Type extends { guid: string }, Initializer> extends Even
...
 
 export type DispatcherScope = Dispatcher<any, any>;
-class Root extends Dispatcher<{ guid: '' }, {}> {
-  constructor(connection: DispatcherConnection) {
+export class Root extends Dispatcher<{ guid: '' }, {}> {
+  constructor(connection: DispatcherConnection, private readonly createPlaywright?: (scope: DispatcherScope) => Promise<PlaywrightDispatcher>) {
     super(connection, { guid: '' }, '', {}, true);
   }
+
+  async initialize(params: { language?: string }): Promise<channels.RootInitializeResult> {
```
Comment (mxschmitt) [inline src/dispatchers/dispatcher.ts:130]: note: will add the language wiring for the error message in a follow-up PR.

File: src/client/channelOwner.ts [hunk 4]
```diff
@@ -78,7 +78,7 @@ export abstract class ChannelOwner<T extends channels.Channel = channels.Channel
         if (prop === 'debugScopeState')
           return (params: any) => this._connection.sendMessageToServer(this, prop, params, stackTrace);
         if (typeof prop === 'string') {
-          const validator = scheme[paramsName(this._type, prop)];
+          const validator = scheme[paramsName(this._type === '' ? 'Root' : this._type, prop)];
```
Comment (dgozman) [inline src/client/channelOwner.ts:81]: Why don't we change Root constructor to pass 'Root' type instead?

File: src/dispatchers/dispatcher.ts [hunk 5]
```diff
@@ -121,15 +122,21 @@ export class Dispatcher<Type extends { guid: string }, Initializer> extends Even
...
+    assert(this.createPlaywright);
+    return {
+      playwright: await this.createPlaywright(this),
```
Comment (dgozman) [inline src/dispatchers/dispatcher.ts:133]: - Let's throw if this method is called twice.
- Why pass `scope` if everyone already has access to the `Root` anyway?

File: src/dispatchers/dispatcher.ts [hunk 6]
```diff
@@ -175,6 +180,8 @@ export class DispatcherConnection {
     };
     const scheme = createScheme(tChannel);
     this._validateParams = (type: string, method: string, params: any): any => {
+      if (type === '')
```
Comment (dgozman) [inline src/dispatchers/dispatcher.ts:183]: Similarly, make it a have a type 'Root' in the constructor?

**Review by dgozman — APPROVED** [review]
This looks pretty good, just a few minor comments.

File: src/client/channelOwner.ts (hunk 4 above)
Comment (mxschmitt) [reply src/client/channelOwner.ts:81]: yeah was a hacky workaround, calling it actually correctly is way better.

**Pull request merged**